src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -21,8 +30,104 @@ public function getUrl(string $bucket, string $key): string
         return $this->getEndpoint($uri, [], null);
     }
 
-    /*
-     * TODO Add get() and put() which supports large files
-     * TODO add has()
+    public function download(string $bucket, string $key): ResultStream
+    {
+        return $this->getObject(['Bucket' => $bucket, 'Key' => $key])->getBody();
+    }
+
+    public function has(string $bucket, string $key): bool
+    {
+        return $this->objectExists(['Bucket' => $bucket, 'Key' => $key])->isSuccess();
+    }
+
+    /**
+     * @param string|resource|callable|iterable $object
+     * @param array{
+     *   ACL?: \AsyncAws\S3\Enum\ObjectCannedACL::*,
+     *   CacheControl?: string,
+     *   ContentType?: string,
+     *   Metadata?: array<string, string>,
+     * } $options
      */
+    public function upload(string $bucket, string $key, $object, array $options = []): void
+    {
+        // Send upload requests with size 32 MB
+        $megabyte = 1024 * 1024;
+        $stream = $this->getStream($object, 32 * $megabyte);
```

Comment:
what's about...
```php
$partSize = max(64, ceil($contentLength / (10000 * 1024 * 1024))); // number of chunk per part
$stream = $this->getStream($object, 1 * 1024 * 1024); // split the stream in 1MB chunk

$buffer = \fopen('php://temp', 'rw+');
$chunkIndex = 0;
foreach ($stream as$chunk) {
  if (null === $buffer) {
  }
  fwrite($buffer, $chunk);
  if (++$chunkIndex >= $partSize) {
   $response = $this->uploadPart(['Body' => $buffer]);
   $buffer = \fopen('php://temp', 'rw+');
   $chunkIndex = 0;
  }
}
```

Reply:
this will handle all cases (froms few MB to several TB) with less than 2Mb memory footprint.
Whithout too much complexity

Reply:
Yes, that would take less memory. I'll implement something like it.